File: max_api_bridge.js
#!/usr/bin/env node
/**
 * Max for Live <-> Flask API Bridge (Node.js version)
 * Accepts commands via Max's node.script, sends HTTP requests, returns JSON
 * This allows Max to communicate with the Flask API server
 */

const http = require('http');
const Max = require('max-api');

const API_HOST = 'localhost';
const API_PORT = 8080;

/**
 * Make HTTP request to Flask server
 */
function makeRequest(method, path, data = null) {
    return new Promise((resolve, reject) => {
        // Prepare request body if data is provided
        const requestBody = data ? JSON.stringify(data) : null;

        const options = {
            hostname: API_HOST,
            port: API_PORT,
            path: path,
            method: method,
            headers: {
                'Content-Type': 'application/json',
            }
        };

        // Add Content-Length header for POST/PUT requests
        if (requestBody) {
            options.headers['Content-Length'] = Buffer.byteLength(requestBody);
        }

        const req = http.request(options, (res) => {
            let body = '';

            res.on('data', (chunk) => {
                body += chunk;
            });

            res.on('end', () => {
                try {
                    const jsonData = JSON.parse(body);
                    resolve({ status: res.statusCode, data: jsonData });
                } catch (e) {
                    resolve({ status: res.statusCode, data: body });
                }
            });
        });

        req.on('error', (error) => {
            reject(error);
        });

        // Send data if provided
        if (requestBody) {
            req.write(requestBody);
        }

        req.end();
    });
}

/**
 * Check if API server is running
 */
async function checkHealth() {
    try {
        const result = await makeRequest('GET', '/health');

        if (result.status === 200) {
            return {
                status: 'success',
                server: 'healthy',
                model_loaded: result.data.model_loaded || false
            };
        } else {
            return {
                status: 'error',
                message: `Server returned ${result.status}`
            };
        }
    } catch (error) {
        return {
            status: 'error',
            message: error.message
        };
    }
}

/**
 * Generate preset from text description
 */
async function generatePreset(description) {
    try {
        const result = await makeRequest('POST', '/generate', { description });

        if (result.status === 200) {
            return {
                status: 'success',
                preset: result.data
            };
        } else {
            return {
                status: 'error',
                message: `Server returned ${result.status}`
            };
        }
    } catch (error) {
        return {
            status: 'error',
            message: error.message
        };
    }
}

/**
 * Handle incoming messages from Max
 */
Max.addHandler('health', async () => {
    const result = await checkHealth();
    Max.outlet(JSON.stringify(result));
});

Max.addHandler('generate', async (requestJson) => {
    try {
        // Parse the JSON request from Max
        Max.post(`[Bridge] Received generate request: ${requestJson}`);
        const request = JSON.parse(requestJson);
        const description = request.description || '';
        Max.post(`[Bridge] Description: ${description}`);

        const result = await generatePreset(description);
        Max.post(`[Bridge] Result status: ${result.status}`);
        Max.outlet(JSON.stringify(result));
    } catch (error) {
        Max.post(`[Bridge] Error: ${error.message}`);
        Max.outlet(JSON.stringify({
            status: 'error',
            message: `Failed to parse request: ${error.message}`
        }));
    }
});

// Startup message
Max.post('M1 API Bridge (Node.js) ready');
Max.post(`Connecting to Flask server at ${API_HOST}:${API_PORT}`);

File: old/max_api_bridge.py
#!/usr/bin/env python3
"""
Max for Live <-> Flask API Bridge
Accepts commands via stdin, sends HTTP requests, returns JSON via stdout
This allows Max to communicate with the Flask API server
"""
import sys
import json
import requests

API_URL = "http://localhost:8080"

def check_health():
    """Check if API server is running"""
    try:
        response = requests.get(f"{API_URL}/health", timeout=2)
        if response.status_code == 200:
            data = response.json()
            return {"status": "success", "server": "healthy", "model_loaded": data.get("model_loaded", False)}
        else:
            return {"status": "error", "message": f"Server returned {response.status_code}"}
    except requests.exceptions.RequestException as e:
        return {"status": "error", "message": str(e)}

def generate_preset(description):
    """Generate preset from text description"""
    try:
        response = requests.post(
            f"{API_URL}/generate",
            json={"description": description},
            timeout=60  # AI generation can take 20-30 seconds
        )

        if response.status_code == 200:
            preset_data = response.json()
            return {"status": "success", "preset": preset_data}
        else:
            return {"status": "error", "message": f"Server returned {response.status_code}"}
    except requests.exceptions.RequestException as e:
        return {"status": "error", "message": str(e)}

def main():
    """Main loop - read commands from stdin, write responses to stdout"""
    # Read command from stdin
    for line in sys.stdin:
        try:
            command = json.loads(line.strip())
            action = command.get("action")

            if action == "health":
                result = check_health()
            elif action == "generate":
                description = command.get("description", "")
                result = generate_preset(description)
            else:
                result = {"status": "error", "message": f"Unknown action: {action}"}

            # Write result to stdout (Max will read this)
            print(json.dumps(result), flush=True)

        except json.JSONDecodeError as e:
            print(json.dumps({"status": "error", "message": f"Invalid JSON: {e}"}), flush=True)
        except Exception as e:
            print(json.dumps({"status": "error", "message": str(e)}), flush=True)

if __name__ == "__main__":
    main()
